### PR TITLE
[ Transpose sinking ] Transpose->FQ->Reduce

### PR DESCRIPTION
**Pull request by jane-intel**
### Details:
 - Transpose->FQ->Reduce to FQ->Reduce->Transpose
 - Enabled RTTI for Reduction classes.

image-2021-04-01-00-27-32-120

### Tickets:
 - *43189*

### CODE REVIEW

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+ngraph::pass::TransposeReduction::TransposeReduction() {
+    MATCHER_SCOPE(TransposeReduction);
+
+    auto transpose_label = pattern::wrap_type<opset6::Transpose>({pattern::any_input(), pattern::wrap_type<opset6::Constant>()});
+    auto reduce_or_squeeze_label = pattern::wrap_type<op::util::ArithmeticReductionKeepDims, op::util::LogicalReductionKeepDims, opset6::Squeeze>(
+            {transpose_label, pattern::wrap_type<opset6::Constant>()});
+
+    ngraph::matcher_pass_callback matcher_pass_callback = [=](ngraph::pattern::Matcher &m) {
+        auto &pattern_to_output = m.get_pattern_value_map();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp:167]: nit pick
```suggestion
        const auto &pattern_to_output = m.get_pattern_value_map();
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+
+bool replace_transpose_with_reshape(const std::shared_ptr<Node>& transpose) {
+    auto data = transpose->input_value(0);
+    const auto input_shape = transpose->input(0).get_partial_shape();
+
+    const size_t input_shape_rank = input_shape.rank().get_length();
+
+    auto order = as_type_ptr<opset3::Constant>(transpose->input_value(1).get_node_shared_ptr());
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp:65]: Please use the same opset6

**Review by GlebKazantaev — APPROVED** [review]

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp (hunk 2 above)
Comment (jane-intel) [inline inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp:65]:
```suggestion
    auto order = as_type_ptr<opset6::Constant>(transpose->input_value(1).get_node_shared_ptr());
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    MATCHER_SCOPE(TransposeOptimization);
+
+    auto transpose_label = pattern::wrap_type<opset6::Transpose>(
+            {pattern::any_input(pattern::has_static_rank()), pattern::wrap_type<opset6::Constant>()});
+    ngraph::matcher_pass_callback matcher_pass_callback = [=](ngraph::pattern::Matcher &m) {
+        auto &pattern_to_output = m.get_pattern_value_map();
+        auto transpose = as_type_ptr<opset6::Transpose>(pattern_to_output.at(transpose_label).get_node_shared_ptr());
+        if (!transpose)
+            return false;
+        return replace_transpose_with_reshape(transpose);
```
Comment (jane-intel) [inline inference-engine/src/transformations/src/transformations/common_optimizations/transpose_sinking.cpp:152]:
```suggestion
    ngraph::matcher_pass_callback matcher_pass_callback = [=](ngraph::pattern::Matcher &m) {
        return replace_transpose_with_reshape(m.get_match_root());
```

**Pull request merged**